File: DemoServer/AdditionalLanguages/python/multiMapIndexes/multiMapReduceIndex/multiMapReduceIndex.py
from demo_example import Example, RunParamsBase

#region Usings
from ravendb.documents.indexes.index_creation import AbstractMultiMapIndexCreationTask
#endregion

class RunParams(RunParamsBase):
    def __init__(self, min_companies_count: int, min_items_count: int):
        self.min_companies_count = min_companies_count
        self.min_items_count = min_items_count

#region Demo
#region Step_1
class CityCommerceDetails(AbstractMultiMapIndexCreationTask):
#endregion

    #region Step_2
    class IndexEntry:
        def __init__(
            self,
            city_name: str = None,
            number_of_companies_in_city: int = None,
            number_of_suppliers_in_city: int = None,
            number_of_items_shipped_to_city: int = None,
        ):
            self.city_name = city_name
            self.number_of_companies_in_city = number_of_companies_in_city
            self.number_of_suppliers_in_city = number_of_suppliers_in_city
            self.number_of_items_shipped_to_city = number_of_items_shipped_to_city
    #endregion

    def __init__(self):
        #region Step_3
        super().__init__()
        self._add_map(
            (
                "docs.Companies.Select(company => new {"
                "    city_name = company.Address.City,"
                "    number_of_companies_in_city = 1,"
                "    number_of_suppliers_in_city = 0,"
                "    number_of_items_shipped_to_city = 0"
                "})"
            )
        )

        self._add_map(
            (
                "docs.Suppliers.Select(supplier => new {"
                "    city_name = supplier.Address.City,"
                "    number_of_companies_in_city = 0,"
                "    number_of_suppliers_in_city = 1,"
                "    number_of_items_shipped_to_city = 0"
                "})"
            )
        )

        self._add_map(
            (
                "docs.Orders.Select(order => new {"
                "    city_name = order.ShipTo.City,"
                "    number_of_companies_in_city = 0,"
                "    number_of_suppliers_in_city = 0,"
                "    number_of_items_shipped_to_city = Enumerable.Sum(order.Lines, x => ((int) x.Quantity))"
                "})"
            )
        )
        #endregion
        
        #region Step_4
        self.reduce = (
            "results.GroupBy(result => result.city_name).Select(g => new {"
            "    city_name = g.Key,"
            "    number_of_companies_in_city = Enumerable.Sum(g, x => ((int) x.number_of_companies_in_city)),"
            "    number_of_suppliers_in_city = Enumerable.Sum(g, x0 => ((int) x0.number_of_suppliers_in_city)),"
            "    number_of_items_shipped_to_city = Enumerable.Sum(g, x1 => ((int) x1.number_of_items_shipped_to_city))"
            "})"
        )
        #endregion
#endregion

class MultiMapReduceIndex(Example):
    def run(self, run_params: RunParams):
        minCompaniesCount = run_params.min_companies_count
        minItemsCount = run_params.min_items_count
        CityCommerceDetails().execute(self.document_store_holder.store())
        #region Demo
        #region Step_5
        with self.document_store_holder.store().open_session() as session:
            commerce_details = list(
                session.query_index_type(CityCommerceDetails, CityCommerceDetails.IndexEntry)
                .where_greater_than("number_of_companies_in_city", minCompaniesCount)
                .or_else()
                .where_greater_than("number_of_items_shipped_to_city", minItemsCount)
                .order_by("city_name")
            )
        #endregion
        #endregion

        return commerce_details

File: DemoServer/AdditionalLanguages/php/multiMapIndexes/multiMapReduceIndex/MultiMapReduceIndex.php
<?php

namespace RavenDB\Demo\multiMapIndex\multiMapReduceIndex;

//region Usings
use RavenDB\Documents\Indexes\AbstractMultiMapIndexCreationTask;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;

//region demo
//region Step_1
class CityCommerceDetails extends AbstractMultiMapIndexCreationTask
//endregion
{
    public function __construct()
    {
        parent::__construct();

        //region Step_2
        $this->addMap("docs.Companies.Select(company => new {" .
            "    CityName = company.Address.City," .
            "    NumberOfCompaniesInCity = 1," .
            "    NumberOfSuppliersInCity = 0," .
            "    NumberOfItemsShippedToCity = 0" .
            "})");

        $this->addMap("docs.Suppliers.Select(supplier => new {" .
            "    CityName = supplier.Address.City," .
            "    NumberOfCompaniesInCity = 0," .
            "    NumberOfSuppliersInCity = 1," .
            "    NumberOfItemsShippedToCity = 0" .
            "})");

        $this->addMap("docs.Orders.Select(order => new {" .
            "    CityName = order.ShipTo.City," .
            "    NumberOfCompaniesInCity = 0," .
            "    NumberOfSuppliersInCity = 0," .
            "    NumberOfItemsShippedToCity = Enumerable.Sum(order.Lines, x => ((int) x.Quantity))" .
            "})");
        //endregion
        
        //region Step_3
        $this->reduce = "results.GroupBy(result => result.CityName).Select(g => new {" .
            "    CityName = g.Key," .
            "    NumberOfCompaniesInCity = Enumerable.Sum(g, x => ((int) x.NumberOfCompaniesInCity))," .
            "    NumberOfSuppliersInCity = Enumerable.Sum(g, x0 => ((int) x0.NumberOfSuppliersInCity))," .
            "    NumberOfItemsShippedToCity = Enumerable.Sum(g, x1 => ((int) x1.NumberOfItemsShippedToCity))" .
            "})";
        //endregion
    }

}
//endregion

//region demo
//region Step_4
class IndexEntry
{
    private ?string $cityName = null;
    private ?int $numberOfCompaniesInCity = null;
    private ?int $numberOfSuppliersInCity = null;
    private ?int $numberOfItemsShippedToCity = null;

    public function getCityName(): ?string
    {
        return $this->cityName;
    }

    public function setCityName(?string $cityName): void
    {
        $this->cityName = $cityName;
    }

    public function getNumberOfCompaniesInCity(): ?int
    {
        return $this->numberOfCompaniesInCity;
    }

    public function setNumberOfCompaniesInCity(?int $numberOfCompaniesInCity): void
    {
        $this->numberOfCompaniesInCity = $numberOfCompaniesInCity;
    }

    public function getNumberOfSuppliersInCity(): ?int
    {
        return $this->numberOfSuppliersInCity;
    }

    public function setNumberOfSuppliersInCity(?int $numberOfSuppliersInCity): void
    {
        $this->numberOfSuppliersInCity = $numberOfSuppliersInCity;
    }

    public function getNumberOfItemsShippedToCity(): ?int
    {
        return $this->numberOfItemsShippedToCity;
    }

    public function setNumberOfItemsShippedToCity(?int $numberOfItemsShippedToCity): void
    {
        $this->numberOfItemsShippedToCity = $numberOfItemsShippedToCity;
    }
}
//endregion
//endregion

class MultiMapReduceIndex
{
    public function __invoke(RunParams $runParams): array
    {
        $minCompaniesCount = $runParams->getMinCompaniesCount() ?? 5;
        $minItemsCount = $runParams->getMinItemsCount() ?? 2000;
        
        //region Demo
        //region Step_5
        $commerceDetails = [];

        $session = DocumentStoreHolder::getStore()->openSession();
        
        try {
            $commerceDetails = $session->query(IndexEntry::class, CityCommerceDetails::class)
                ->whereGreaterThan("NumberOfCompaniesInCity", $minCompaniesCount)
                ->orElse()
                ->whereGreaterThan("NumberOfItemsShippedToCity", $minItemsCount)
                ->orderBy("CityName")
                ->toList();
                
        } finally {
            $session->close();
        }
        //endregion
        //endregion

        return $commerceDetails;
    }
}
